File: android/app/src/main/java/com/echo/dictation/speech/provider/ProviderRegistry.kt
package com.echo.dictation.speech.provider

/**
 * Central registry of all supported speech providers.
 *
 * Adding a new provider is a two-step change:
 *   1. Add an entry to [ProviderId] enum in SpeechProvider.kt.
 *   2. Add a [ProviderConfig] entry here.
 *
 * No UI code or repository code needs to change.
 */
object ProviderRegistry {

    val allConfigs: List<ProviderConfig> = listOf(
        ProviderConfig(
            id             = ProviderId.GROQ,
            displayName    = "Groq",
            defaultBaseUrl = "https://api.groq.com/openai/v1/",
            models         = listOf("whisper-large-v3-turbo", "whisper-large-v3"),
            authHeaderName = "Authorization",
            authValueFormat = "Bearer %s",
        ),
        ProviderConfig(
            id             = ProviderId.OPENAI,
            displayName    = "OpenAI",
            defaultBaseUrl = "https://api.openai.com/v1/",
            models         = listOf("whisper-1", "gpt-4o-transcribe", "gpt-4o-mini-transcribe"),
            authHeaderName = "Authorization",
            authValueFormat = "Bearer %s",
        ),
        ProviderConfig(
            id             = ProviderId.OPENROUTER,
            displayName    = "OpenRouter",
            defaultBaseUrl = "https://openrouter.ai/api/v1/",
            models         = listOf("openai/whisper-large-v3", "openai/whisper"),
            authHeaderName = "Authorization",
            authValueFormat = "Bearer %s",
        ),
        ProviderConfig(
            id             = ProviderId.DEEPGRAM,
            displayName    = "Deepgram",
            defaultBaseUrl = "https://api.deepgram.com/",
            models         = listOf("nova-3", "nova-2"),
            authHeaderName = "Authorization",
            authValueFormat = "Token %s",
        ),
        ProviderConfig(
            id             = ProviderId.ASSEMBLYAI,
            displayName    = "AssemblyAI",
            defaultBaseUrl = "https://api.assemblyai.com/",
            models         = listOf("default"),
            authHeaderName = "Authorization",
            authValueFormat = "%s",           // no "Bearer" prefix for AssemblyAI
        ),
        ProviderConfig(
            id             = ProviderId.GEMINI,
            displayName    = "Google Gemini",
            defaultBaseUrl = "https://generativelanguage.googleapis.com/",
            models         = listOf("gemini-2.0-flash", "gemini-1.5-flash"),
            authHeaderName = "x-goog-api-key",
            authValueFormat = "%s",           // header value IS the key directly
        ),
        ProviderConfig(
            id                    = ProviderId.AZURE,
            displayName           = "Azure OpenAI",
            defaultBaseUrl        = "",        // user must supply the deployment endpoint
            models                = emptyList(),
            requiresCustomBaseUrl = true,
            requiresCustomModel   = true,
            authHeaderName        = "api-key",
            authValueFormat       = "%s",
        ),
        ProviderConfig(
            id                    = ProviderId.BEDROCK,
            displayName           = "AWS Bedrock",
            // Base URL is region-specific: https://bedrock-runtime.<region>.amazonaws.com/
            // The user must supply their region endpoint in the custom URL field.
            defaultBaseUrl        = "",
            models                = listOf(
                "amazon.nova-lite-v1:0",
                "amazon.nova-pro-v1:0",
                "anthropic.claude-3-5-sonnet-20241022-v2:0",
                "anthropic.claude-3-7-sonnet-20250219-v1:0",
                "anthropic.claude-sonnet-4-5",
                "anthropic.claude-opus-4-5",
                "meta.llama3-3-70b-instruct-v1:0",
                "mistral.mistral-large-2402-v1:0",
            ),
            requiresCustomBaseUrl = true,
            requiresCustomModel   = false,
            // Bedrock uses AWS SigV4 auth. The API key field stores
            // "ACCESS_KEY_ID:SECRET_ACCESS_KEY" and the provider handles signing.
            authHeaderName        = "Authorization",
            authValueFormat       = "Bearer %s",
        ),
        ProviderConfig(
            id                    = ProviderId.CUSTOM,
            displayName          = "Custom OpenAI-Compatible",
            defaultBaseUrl       = "",        // user must supply
            models               = emptyList(),
            requiresCustomBaseUrl = true,
            requiresCustomModel  = true,
            authHeaderName       = "Authorization",
            authValueFormat      = "Bearer %s",
        ),
    )

    /** Convenience map for O(1) lookup. */
    private val configById: Map<ProviderId, ProviderConfig> =
        allConfigs.associateBy { it.id }

    fun getConfig(id: ProviderId): ProviderConfig =
        configById[id] ?: error("No config registered for provider: $id")
}

File: EchoCore/Sources/EchoCore/Services/Providers/ProviderRegistry.swift
//
//  ProviderRegistry.swift
//  EchoCore
//
//  Single source of truth for all provider configurations.
//  V3: LLM chat models and chatBaseURL added to all providers that support text completion.
//

import Foundation

public enum ProviderRegistry {

    // MARK: - All provider configurations

    public static let allConfigs: [ProviderConfig] = [
        ProviderConfig(
            id: .groq,
            displayName: "Groq",
            defaultBaseURL: "https://api.groq.com/openai/v1/",
            supportedModels: ["whisper-large-v3-turbo", "whisper-large-v3"],
            defaultModel: "whisper-large-v3-turbo",
            authHeaderName: "Authorization",
            authHeaderValueFormat: "Bearer %@",
            capabilities: .openAICompatibleLLM,
            chatBaseURL: "https://api.groq.com/openai/v1/",
            defaultChatModel: "llama-3.3-70b-versatile",
            supportedChatModels: [
                "llama-3.3-70b-versatile",
                "llama-3.1-70b-versatile",
                "llama-3.1-8b-instant",
                "mixtral-8x7b-32768",
            ]
        ),
        ProviderConfig(
            id: .openAI,
            displayName: "OpenAI",
            defaultBaseURL: "https://api.openai.com/v1/",
            supportedModels: ["whisper-1", "gpt-4o-transcribe", "gpt-4o-mini-transcribe"],
            defaultModel: "whisper-1",
            authHeaderName: "Authorization",
            authHeaderValueFormat: "Bearer %@",
            capabilities: .openAICompatibleLLM,
            chatBaseURL: "https://api.openai.com/v1/",
            defaultChatModel: "gpt-4o-mini",
            supportedChatModels: ["gpt-4o", "gpt-4o-mini", "gpt-4-turbo", "gpt-3.5-turbo"]
        ),
        ProviderConfig(
            id: .openRouter,
            displayName: "OpenRouter",
            defaultBaseURL: "https://openrouter.ai/api/v1/",
            supportedModels: ["openai/whisper-large-v3", "openai/whisper"],
            defaultModel: "openai/whisper-large-v3",
            authHeaderName: "Authorization",
            authHeaderValueFormat: "Bearer %@",
            capabilities: .openAICompatibleLLM,
            chatBaseURL: "https://openrouter.ai/api/v1/",
            defaultChatModel: "openai/gpt-4o-mini",
            supportedChatModels: [
                "openai/gpt-4o", "openai/gpt-4o-mini",
                "google/gemini-pro", "anthropic/claude-3-haiku",
                "meta-llama/llama-3-70b-instruct",
            ]
        ),
        ProviderConfig(
            id: .deepgram,
            displayName: "Deepgram",
            defaultBaseURL: "https://api.deepgram.com/",
            supportedModels: ["nova-3", "nova-2"],
            defaultModel: "nova-3",
            authHeaderName: "Authorization",
            authHeaderValueFormat: "Token %@",
            capabilities: .deepgram
            // Deepgram has no LLM chat capability
        ),
        ProviderConfig(
            id: .assemblyAI,
            displayName: "AssemblyAI",
            defaultBaseURL: "https://api.assemblyai.com/",
            supportedModels: ["default"],
            defaultModel: "default",
            authHeaderName: "Authorization",
            authHeaderValueFormat: "%@",
            capabilities: .assemblyAI
            // AssemblyAI has no LLM chat capability
        ),
        ProviderConfig(
            id: .gemini,
            displayName: "Google Gemini",
            defaultBaseURL: "https://generativelanguage.googleapis.com/",
            supportedModels: ["gemini-2.0-flash", "gemini-1.5-flash"],
            defaultModel: "gemini-2.0-flash",
            authHeaderName: "x-goog-api-key",
            authHeaderValueFormat: "%@",
            capabilities: .geminiLLM,
            chatBaseURL: "https://generativelanguage.googleapis.com/",
            defaultChatModel: "gemini-1.5-flash",
            supportedChatModels: ["gemini-1.5-pro", "gemini-1.5-flash", "gemini-2.0-flash"]
        ),
        ProviderConfig(
            id: .azure,
            displayName: "Azure OpenAI",
            defaultBaseURL: "",
            supportedModels: [],
            defaultModel: "",
            authHeaderName: "api-key",
            authHeaderValueFormat: "%@",
            capabilities: .azure,
            chatBaseURL: nil,          // user-configurable, same as speech base URL
            defaultChatModel: nil,
            supportedChatModels: []
        ),
        ProviderConfig(
            id: .custom,
            displayName: "Custom OpenAI-Compatible",
            defaultBaseURL: "",
            supportedModels: [],
            defaultModel: "",
            authHeaderName: "Authorization",
            authHeaderValueFormat: "Bearer %@",
            capabilities: .custom,
            chatBaseURL: nil,
            defaultChatModel: nil,
            supportedChatModels: []
        ),
    ]

    // MARK: - Lookups

    public static func configuration(for id: ProviderId) -> ProviderConfig {
        guard let config = allConfigs.first(where: { $0.id == id }) else {
            preconditionFailure("No configuration registered for \(id.rawValue)")
        }
        return config
    }

    public static func configuration(forRawValue rawValue: String) -> ProviderConfig? {
        guard let id = ProviderId(rawValue: rawValue) else { return nil }
        return allConfigs.first { $0.id == id }
    }

    // MARK: - Capability presets

    // Speech-only (no LLM)
    public static let openAICompatibleCapabilities = ProviderCapabilities(
        supportsStreaming: false,
        supportsLanguageSelection: true,
        supportsPrompt: true,
        supportsVerboseJSON: true,
        supportsWordTimestamps: false,
        supportsTemperature: true,
        supportsCustomBaseURL: false,
        supportsCustomModel: false,
        supportsConnectionTest: true,
        supportsTextCompletion: false,
        supportsSystemPrompt: false
    )

    // Speech + LLM (Groq, OpenAI, OpenRouter)
    public static let openAICompatibleLLMCapabilities = ProviderCapabilities(
        supportsStreaming: false,
        supportsLanguageSelection: true,
        supportsPrompt: true,
        supportsVerboseJSON: true,
        supportsWordTimestamps: false,
        supportsTemperature: true,
        supportsCustomBaseURL: false,
        supportsCustomModel: false,
        supportsConnectionTest: true,
        supportsTextCompletion: true,
        supportsSystemPrompt: true
    )

    public static let deepgramCapabilities = ProviderCapabilities(
        supportsLanguageSelection: true,
        supportsConnectionTest: true
    )

    public static let assemblyAICapabilities = ProviderCapabilities(
        supportsLanguageSelection: true,
        supportsConnectionTest: true
    )

    // Gemini speech only
    public static let geminiCapabilities = ProviderCapabilities(
        supportsLanguageSelection: true,
        supportsPrompt: true,
        supportsConnectionTest: true
    )

    // Gemini with LLM
    public static let geminiLLMCapabilities = ProviderCapabilities(
        supportsLanguageSelection: true,
        supportsPrompt: true,
        supportsConnectionTest: true,
        supportsTextCompletion: true,
        supportsSystemPrompt: true
    )

    public static let azureCapabilities = ProviderCapabilities(
        supportsLanguageSelection: true,
        supportsPrompt: true,
        supportsVerboseJSON: true,
        supportsTemperature: true,
        supportsCustomBaseURL: true,
        supportsCustomModel: true,
        supportsConnectionTest: true,
        supportsTextCompletion: true,
        supportsSystemPrompt: true
    )

    public static let customCapabilities = azureCapabilities
}

// MARK: - Convenience short-hands used inside ProviderConfig initialisers

private extension ProviderCapabilities {
    static let openAICompatible    = ProviderRegistry.openAICompatibleCapabilities
    static let openAICompatibleLLM = ProviderRegistry.openAICompatibleLLMCapabilities
    static let deepgram            = ProviderRegistry.deepgramCapabilities
    static let assemblyAI          = ProviderRegistry.assemblyAICapabilities
    static let gemini              = ProviderRegistry.geminiCapabilities
    static let geminiLLM           = ProviderRegistry.geminiLLMCapabilities
    static let azure               = ProviderRegistry.azureCapabilities
    static let custom              = ProviderRegistry.customCapabilities
}
